flake.nix

```diff
@@ -18,51 +18,62 @@
       let
         pkgs = nixpkgs.legacyPackages.${system};
         poetry2nixPkg = poetry2nix.legacyPackages.${system};
+
+        controllerOverrides = final: prev:
+          let
+            addNativeBuildInputs = drvName: inputs: {
+              "${drvName}" = prev.${drvName}.overridePythonAttrs (
+                old: {
+                  nativeBuildInputs =
+                    (old.nativeBuildInputs or [ ]) ++ inputs;
+                }
+              );
+            };
+          in
+          { }
+          // addNativeBuildInputs "mpi4py" [ final.cython ]
+          // addNativeBuildInputs "rexfw" [ final.poetry-core ]
+          // addNativeBuildInputs "chainsail-helpers" [ final.poetry-core ];
+
+        controllerOpts = {
+          projectDir = ./app/controller;
+          overrides = poetry2nixPkg.overrides.withDefaults(controllerOverrides);
+          preferWheels = true;
+        };
+        controller = poetry2nixPkg.mkPoetryApplication controllerOpts;
+        controllerEnv = poetry2nixPkg.mkPoetryEnv controllerOpts;
+
+        basePackages = [
+          pkgs.docker-compose
+          pkgs.file
+          pkgs.gnumake
+          pkgs.kubectl
+          pkgs.kubernetes-helm
+          pkgs.minikube
+          pkgs.ncurses
+          pkgs.nodejs
+          pkgs.openmpi
+          pkgs.poetry
+          pkgs.python38Packages.tkinter
+          pkgs.terraform
+          pkgs.yarn
+        ];
+
+        pythonDevShell = ps: pkgs.mkShell {
+          packages = ps ++ basePackages;
+        };
+        controllerDevShell = pythonDevShell [
```

Comment:
I now also tested your proposed developer workflow and that works smoothly! Yes, it's a couple of steps, but I don't think there is a way around this when working fully with Nix. It would be cool, though, if we could replace the `python -m chainsail.controller.run_local` incantation with something like an alias `run-controller-local`, similar to the Poetry script when you use the Poetry shell. Maybe there is an option to pass through a `shellHook` in `mkPoetryEnv`, but a quick search was inconclusive. But that shouldn't prevent merging, just a nice to have :slightly_smiling_face:

Reply:
It would be straightforward to add a `shellHook` if you're interested. Not to `mkPoetryEnv`, but just in the usual place as an argument to `mkShell`.

Reply:
True, in `mkShell` it's easy! Let's do that then; it would then somewhat maintain the interface of the previous way of running the controller locally.

Reply:
@steshaw If you'd like to add the above `mkShell` argument, then we can merge this, I think. But even without it should be fine, but in any case we should document the Nix-based development workflow.

Reply:
Thanks for the bump here, I'd forgotten about adding an alias or script to the development environment. I'd really like to get this merged. Unfortunately, this PR is stacked on top of https://github.com/tweag/chainsail/pull/446 which is blocked by https://github.com/tweag/chainsail/issues/449. However, I was able to test an end-to-end workflow for #446 (which I assume exercises Celery tasks). To start the frontend dev server for the E2E workflow, I temporarily downgraded the nixpkgs pin to `nixpkgs/21.11`. I was able to create, start, and finish a job using the "mixture.zip". I'll make a similar note against #446 in case you are happy to merge that given that.

Reply:
It turns out that there is no need for an alias or script as the environment already includes the scripts from `pyproject.toml`!

```sh-session
$ which chainsail-controller-local
/nix/store/ax7z9q980m2mxg9s468icylb06285nwr-python3-3.10.9-env/bin/chainsail-controller-local
```

Reply:
Oh nice, I didn't realize these get carried over when using `mkPoetryEnv`!

Reply:
I guess that script is also correctly wrapped and everything by poetry2nix?

Reply:
Yes, exactly